examples/bank_agent_eval.py

```diff
@@ -39,25 +90,83 @@ def main():
 
     schema_description = ba.get_schema_description(db_connection)
 
-    system_message = ba.get_system_message()
+    system_message = ba.get_system_message(
+        datetime.datetime.strptime('2023-08-31', '%Y-%m-%d'))
     system_message += schema_description
 
     query_database, _ = tools.make_sqlite_query_tool(db_connection)
     functions = {
         'query_database': query_database
     }
 
-    # Systems to evaluate. These take a question as input and return
+    def get_table_schema(name: str) -> str:
+        return sqlite_utils.get_table_schema(db_connection, name).to_string()
+
+    def build_query_db_debug(conn):
+        def query_db_debug(query: str) -> str:
+            print(f'\t\tQUERY STRING: ```{query}```')
+            try:
+                res = pd.read_sql_query(query, conn).to_string()
+            except Exception as error:
+                res = f'Query failed: {error}'
+            return res
+        return query_db_debug
+
+    functions_with_get_schema = {
+        # 'query_database': query_database,
+        'query_database': build_query_db_debug(db_connection),
+        'get_table_schema': get_table_schema
+    }
+
+    # Systems to evaluate. These take a question and seed as input and return
     # an answer or None (for an uncaught error or if the system can't
     # answer the question) as well as a count of API / LLM interactions
 
-    def build_agent(model_name: str, simple: bool) -> Callable:
-        """build an agent"""
+    def build_agent_simple_llama2(model_file_name: str, promptformat: Callable, stop: List[str]) -> Callable:
+        """build an agent using a local llama model"""
+
+        def call(question: str, seed: int) -> Tuple[Optional[str], int]:
+            """answer a question with the react agent"""
+
+            model_path = os.path.join(project.MODELS_DIR_PATH, model_file_name)
+            assert os.path.isfile(model_path), 'mmodel file not found'
+
+            llama_model = llama_cpp.Llama(
+                model_path=model_path,
+                n_gpu_layers=10000 if GPU_ENABLE else 0,
+                n_ctx=simplelocal.LLAMA2_MAX_TOKENS,
+                seed=seed  # AFAIK this is the only way to set the seed.
+            )
+
+            llama_model_wrapped = simplelocal.build_llama2_wrapper(
+                llama_model, promptformat, stop)
+
+            question_uid = hash(question) % 1000
```

Comment:
May not matter in this instance, but `hash` is not generally reproducible so the uid will only unique in a given run.  `hashlib` has alternatives.

Not sure what question_uid is being used for, so feel free to ignore if this comment is not relevant.  I imagine if this were an issue you would have noticed already :-)

Reply:
Ah nice catch. I thought "hash" did something sensible for strings but I guess not. Was just using this so the saved prepared prompts were unique. I can replace with something better.